python/lib/core/dmod/core/dataset.py

```diff
@@ -59,85 +59,70 @@ class Dataset(Serializable):
-                       derived_from=json_obj[cls._KEY_DERIVED_FROM] if cls._KEY_DERIVED_FROM in json_obj else None,
-                       derivations=json_obj[cls._KEY_DERIVATIONS] if cls._KEY_DERIVATIONS in json_obj else [],
-                       created_on=cls._date_parse_helper(json_obj, cls._KEY_CREATED_ON),
-                       last_updated=cls._date_parse_helper(json_obj, cls._KEY_LAST_UPDATE))
-        except Exception as e:
+    _SERIAL_DATETIME_STR_FORMAT: ClassVar = '%Y-%m-%d %H:%M:%S'
+    name: str = Field(description="The name for this dataset, which also should be a unique identifier.")
+    category: DataCategory = Field(None, alias="data_category", description="The ::class:`DataCategory` type value for this instance.")
+    data_domain: DataDomain
+    dataset_type: DatasetType = Field(DatasetType.UNKNOWN, alias="type")
+    access_location: str = Field(description="String representation of the location at which this dataset is accessible.")
+    uuid: Optional[UUID] = Field(default_factory=uuid4)
+    # manager can only be passed as constructed DatasetManager subtype. Manager not included in `dict` or `json` deserialization.
+    # TODO: don't include `manager` in `Dataset.schema()`. Inclusion is not reflective of the de/serialization behavior.
+    manager: Optional['DatasetManager'] = Field(exclude=True)
+    manager_uuid: Optional[UUID]
+    is_read_only: bool = Field(True, description="Whether this is a dataset that can only be read from.")
+    description: Optional[str]
+    expires: Optional[datetime] = Field(description='The time after which a dataset may "expire" and be removed, or ``None`` if the dataset is not temporary.')
+    derived_from: Optional[str] = Field(description="The name of the dataset from which this dataset was derived, if it is known to have been derived.")
+    derivations: Optional[List[str]] = Field(default_factory=list, description="""List of names of datasets which were derived from this dataset.\n
+    Note that it is not guaranteed that any such dataset still exist and/or are still available.""")
+    created_on: Optional[datetime] = Field(description="When this dataset was created, or ``None`` if that is not known.")
+    last_updated: Optional[datetime]
+
+    @validator("created_on", "last_updated", "expires", pre=True)
+    def parse_dates(cls, v):
+        if v is None:
             return None
 
-    def __eq__(self, other):
-        return isinstance(other, Dataset) and self.name == other.name and self.category == other.category \
-               and self.dataset_type == other.dataset_type and self.data_domain == other.data_domain \
-               and self.access_location == other.access_location and self.is_read_only == other.is_read_only \
-               and self.created_on == other.created_on
+        if isinstance(v, datetime):
+            return v
+
+        return datetime.strptime(v, cls.get_datetime_str_format())
```

Comment:
> Another option is to bring in and use `dateutil.parser`. `dateutil` is already used throughout DMOD and it can parse just about anything you throw at it without a defined format and make them timezone aware.

super neat! I always just bring out the hammer that is `pandas.to_datetime`. @robertbartel, how much do the datetime format constrains matter? Is that something we would want to handle with `dateutil.parser`? Or does having to handle timezones just add undesired complexity? I typically just enforce UTC ISO 8601.

Reply:
Here's an example from dateutil:

```python
from dateutil.parser import parse as parse_date
parse_date("1999-03-12")
# datetime.datetime(1999, 3, 12, 0, 0)
parse_date("3-12-1999")
# datetime.datetime(1999, 3, 12, 0, 0)
parse_date("1999/3/12")
# datetime.datetime(1999, 3, 12, 0, 0)
parse_date("1999/3/12 10:30")
# datetime.datetime(1999, 3, 12, 10, 30)
parse_date("1999/3/12 10:30+0000")
# datetime.datetime(1999, 3, 12, 10, 30, tzinfo=tzutc())
parse_date("1999/3/12 10:30+00:00")
# datetime.datetime(1999, 3, 12, 10, 30, tzinfo=tzutc())
parse_date("1999/3/12T10:30+00:00")
# datetime.datetime(1999, 3, 12, 10, 30, tzinfo=tzutc())
parse_date("1999/3/12T10:30Z")
# datetime.datetime(1999, 3, 12, 10, 30, tzinfo=tzutc())
parse_date("1999/3/12 10:30:14.323-0500")
# datetime.datetime(1999, 3, 12, 10, 30, 14, 323000, tzinfo=tzoffset(None, -18000))
```

Reply:
I don't think it matters terribly, except in a few situations (e.g., the aforementioned _Session_ example).  The main concern is that we can consistently serialize and deserialize back and forth without losing or altering state.  Precision will matter in a few situations.  As long as those are maintained as needed, I don't object to switching implementations.

Reply:
I could see this leading to comparing datetimes with and without timezones which will throw. So for now, I am going to leave the implementation as is. We can and should certainly revisit this in a more constrained context in the future.